test/partition-healing-tests.js

```diff
@@ -0,0 +1,354 @@
+                // Also verify it didn't have information about b1: that means
+                // sut didn't try to merge partitions yet.
+                verifySuspects(t, tc, [0], [1, 2])
+            ]
+        }
+        )
+     );
+
+// Assign names to nodes and update membership of B from test coordinator.
+function makeAFaultyInB(t, tc) {
+    return function makeAFaultyInB(list, cb) {
+        tc.fakeNodes[1].status = 'alive';
+        tc.fakeNodes[2].status = 'alive';
+
+        var B = tc.getMembership();
+        // According to B, a1 is faulty
+        _.find(B, {port: tc.fakeNodes[0].port}).status = 'faulty';
+        _.find(B, {port: tc.fakeNodes[0].port}).incarnationNumber = 1337;
+
+        tc.fakeNodes[1].membership = _.cloneDeep(B);
+        tc.fakeNodes[2].membership = _.cloneDeep(B);
+
+        cb(list);
+    }
+}
+
+function removeHealPartitionDiscoResponse(t, tc) {
+    return function removeHealPartitionDiscoResponse(list, cb) {
+        var d = _.find(list, { type: events.Types.AdminHealPartitionDisco });
+        if (d === undefined) {
+            cb(null);
+        } else {
+            list.splice(list.indexOf(d), 1);
+            cb(list);
+        }
+    }
+}
+
+// Verify 'nodes' are treated as suspects by sut. Also, verify sut is not
+// sending 'nodes' information that 'nodes_in_other_partition' are alive.
+function verifySuspects(t, tc, nodes, nodes_in_other_partition) {
+    return function verifyBsuspects(list, cb) {
+        // Which addresses should be declared suspects in the same message.
+        var addresses = _.map(nodes, function(ix) {
+            return tc.fakeNodes[ix].getHostPort();
+        });
+
+        var addresses_in_other_partition = _.map(nodes, function(ix) {
+            return tc.fakeNodes[ix].getHostPort();
+        });
+
+
+        var p = _.find(list, {
+            type: events.Types.Ping,
+            direction: 'request'
+        });
+        if (p === undefined) {
+            return cb(null);
+        }
+        var suspects = [];
+        _.forEach(list, function(msg) {
+            _.forEach(safeJSONParse(msg.arg3).changes, function(change) {
+                if (change.status === 'suspect') {
+                    suspects.push(change.address);
+                } else if (change.status == 'alive') {
+                    if (_.find(addresses_in_other_partition, change.address)) {
+                        t.fail(change.source + " disseminated a change about " +
+                                change.address + " being alive. Unexpected " +
+                                "merge of partitions!");
+                    }
+                }
+            });
+            // We found all suspects we were looking for in this message
+            if (_.isEqual(addresses.sort(), suspects.sort())) {
+                t.ok(true, "Found " + nodes + " suspects (" + suspects + ") we were looking for");
+                _.remove(list, msg);
+                return cb(list);
```

Comment:
the early return confuses me a bit, since we are also checking for an Unexpected merge of partitions, shouldn't we check all the ping requests?